test/ngMock/angular-mocksSpec.js

```diff
@@ -614,47 +614,292 @@ describe('ngMock', function() {
+      });
+
+      it('should not throw an exception when passing a specific delay', function() {
+        expect(function() {browser.defer.flush(100);}).not.toThrow();
+      });
+
+      describe('tasks scheduled during flushing', function() {
+        it('should be flushed if they do not exceed the target delay (when no delay specified)',
+          function() {
+            browser.defer(function() {
+              logFn('1')();
+              browser.defer(function() {
+                logFn('3')();
+                browser.defer(logFn('4'), 1);
+              }, 2);
+            }, 1);
+            browser.defer(function() {
+              logFn('2')();
+              browser.defer(logFn('6'), 4);
+            }, 2);
+            browser.defer(logFn('5'), 5);
+
+            browser.defer.flush(0);
+            expect(browser.defer.now).toEqual(0);
+            expect(log).toEqual('');
+
+            browser.defer.flush();
+            expect(browser.defer.now).toEqual(5);
+            expect(log).toEqual('1;2;3;4;5;');
+          }
+        );
+
+        it('should be flushed if they do not exceed the specified delay',
+          function() {
+            browser.defer(function() {
+              logFn('1')();
+              browser.defer(function() {
+                logFn('3')();
+                browser.defer(logFn('4'), 1);
+              }, 2);
+            }, 1);
+            browser.defer(function() {
+              logFn('2')();
+              browser.defer(logFn('6'), 4);
+            }, 2);
+            browser.defer(logFn('5'), 5);
+
+            browser.defer.flush(0);
+            expect(browser.defer.now).toEqual(0);
+            expect(log).toEqual('');
+
+            browser.defer.flush(4);
+            expect(browser.defer.now).toEqual(4);
+            expect(log).toEqual('1;2;3;4;');
+
+            browser.defer.flush(6);
+            expect(browser.defer.now).toEqual(10);
+            expect(log).toEqual('1;2;3;4;5;6;');
+          }
+        );
+      });
     });
 
-    it('should flush delayed', function() {
-      browser.defer(logFn('A'));
-      browser.defer(logFn('B'), 10);
-      browser.defer(logFn('C'), 20);
-      expect(log).toEqual('');
+    describe('defer.cancel', function() {
+      it('should cancel a pending task', function() {
+        var taskId1 = browser.defer(logFn('A'), 100, 'fooType');
+        var taskId2 = browser.defer(logFn('B'), 200);
+
+        expect(log).toBe('');
+        expect(function() {browser.defer.verifyNoPendingTasks('fooType');}).toThrow();
+        expect(function() {browser.defer.verifyNoPendingTasks();}).toThrow();
 
-      expect(browser.defer.now).toEqual(0);
-      browser.defer.flush(0);
-      expect(log).toEqual('A;');
+        browser.defer.cancel(taskId1);
+        expect(function() {browser.defer.verifyNoPendingTasks('fooType');}).not.toThrow();
+        expect(function() {browser.defer.verifyNoPendingTasks();}).toThrow();
```

Comment:
should we check the type of the error that is thrown?

Reply:
This test focuses on `defer.cancel()`. There are tests in `defer.verifyNoPendingTasks()` that have assertions against the actual error message.